### PR TITLE
win_get_url: Rewrite using AnsibleModule

### PR DESCRIPTION
**Pull request by dagwieers**
##### SUMMARY
Simple rewrite using AnsibleModule.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
win_get_url

##### ANSIBLE VERSION
v2.8

### CODE REVIEW

File: lib/ansible/modules/windows/win_get_url.ps1 [hunk 1]
```diff
@@ -76,13 +105,13 @@ Function CheckModified-File($url, $dest, $headers, $credentials, $timeout, $use_
 
         $webLastMod = $webResponse.LastModified
     } Catch [System.Net.WebException] {
-        $result.status_code = $_.Exception.Response.StatusCode
-        Fail-Json -obj $result -message "Error requesting '$url'. $($_.Exception.Message)"
+        $module.Result.status_code = $_.Exception.Response.StatusCode
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_get_url.ps1]: I believe this is where the test issue is occurring, the move to C# means we are no longer using `ConvertTo-Json`. It seems like the PS cmdlet does a bit more to deal with things like PowerShell objects like a `PSParameterizedProperty` which `JavaScriptSerializer` falls flat on. To changing this to `$module.Result.status_code = $_.Exception.Response.StatusCode.ToString()` to ensure we are dealing with a native .NET type. I could be wrong as to whether it's `StatusCode` or maybe `TotalSeconds` used elsewhere so it's a matter of trying out what works. 

This is going to be a slight regression in features but we need `Ansible.Basic` to be compatibility with both native C# and .NET so we cannot use the PS library to do the work for us.

File: lib/ansible/modules/windows/win_get_url.ps1 [hunk 2]
```diff
@@ -129,50 +158,34 @@ Function Download-File($result, $url, $dest, $headers, $credentials, $timeout, $
...
+    if (-not $module.CheckMode) {
         Try {
             $extWebClient.DownloadFile($url, $dest)
         } Catch [System.Net.WebException] {
-            $result.status_code = [int] $_.Exception.Response.StatusCode
-            $result.elapsed = ((Get-Date) - $module_start).TotalSeconds
-            Fail-Json -obj $result -message "Error downloading '$url' to '$dest': $($_.Exception.Message)"
+            $module.Result.status_code = [int] $_.Exception.Response.StatusCode
+            $module.Result.elapsed = ((Get-Date) - $module_start).TotalSeconds
+            $module.FailJson("Error downloading '$url' to '$dest': $($_.Exception.Message)")
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_get_url.ps1]: Add `$_` as a 2nd argument to FailJson

File: lib/ansible/modules/windows/win_get_url.ps1 [hunk 3]
```diff
@@ -129,50 +158,34 @@ Function Download-File($result, $url, $dest, $headers, $credentials, $timeout, $
...
         } Catch {
-            $result.elapsed = ((Get-Date) - $module_start).TotalSeconds
-            Fail-Json -obj $result -message "Unknown error downloading '$url' to '$dest': $($_.Exception.Message)"
+            $module.Result.elapsed = ((Get-Date) - $module_start).TotalSeconds
+            $module.FailJson("Unknown error downloading '$url' to '$dest': $($_.Exception.Message)")
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_get_url.ps1]: Add `$_` as 2nd argument.

File: lib/ansible/modules/windows/win_get_url.ps1 (hunk 1 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_get_url.ps1]: I was suspecting that much, but I wasn't sure whether I would be doing to conversion, or this was something we wanted to fix in `Ansible.Basic`.

That said, we do cast to `int` a little bit below, so I think we can do this here as well. And if this fails, we should be looking into why that is. TBC

File: lib/ansible/modules/windows/win_get_url.ps1 (hunk 3 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_get_url.ps1]: Fixed.

File: lib/ansible/modules/windows/win_get_url.ps1 (hunk 2 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_get_url.ps1]: Fixed.

**Comment by dagwieers** [comment]
@jborean93 I added 2 `FIXME` comments, because I think we should be changing blank exception handling, by something more specific as much as possible. We can make the error-messages also more revealing in those cases.

Also, we have 2 operations in the try-statement, that either should be 2 try-blocks, or (as I suspect) a single try-statement.

Still, I didn't want to make that change in this PR.

**Comment by jborean93** [comment]
@dagwieers before when exceptions and errors would give us an ugly output having blanket exceptions were nice but now it's less so. I do agree that we should try and shy away from blanket exception catches and make things a bit more specific but happy to investigate that in another PR.

**Comment by mattclay** [comment]
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/93264/14/tests

**Comment by jborean93** [comment]
@dagwieers have fixed the circular reference, I found that using the return object of a cmdlet can cause issues like this. In this case `$dest` was the result of `Join-Path` so I just added `.ToString()` to it.

**Comment by dagwieers** [comment]
Hmmm, I wasn't expecting `$dest` to be the problem... Is there no way we can actually report which entry is causing the issue ? This will not be the last time we see this, it'll be a showstopper for people to develop new modules.

**Comment by jborean93** [comment]
Not that I'm aware of which is not to say it isn't possible. The new serializer uses [JavaScriptSerializer](https://docs.microsoft.com/en-us/dotnet/api/system.web.script.serialization.javascriptserializer?view=netframework-4.7.2) which is purely a .NET C# class and doesn't have any special knowledge of PowerShell objects.

I do agree this will be a nice thing to fix and may involve some work with [JavaScriptTypeResolver](https://docs.microsoft.com/en-us/dotnet/api/system.web.script.serialization.javascripttyperesolver?view=netframework-4.7.2) but I haven't used that before.

**Comment by jborean93** [comment]
@dagwieers I've opened a PR https://github.com/ansible/ansible/pull/48860 that removes this problem. It will revert back to using the `ConvertTo-Json` cmdlet when running in a PowerShell runspace (called from PowerShell). This brings back the old serialization behaviour and makes it easier to support PSObjects. I will probably want nitzmahone to have a look when he is ready but from a basic test it works fine for me.

**Pull request merged**